tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -836,17 +835,119 @@ export class WebGPUBackend extends KernelBackend {
+      });
+    } else {
+      bindGroupLayoutEntries.push(
+          {binding: 1, visibility: GPUShaderStage.COMPUTE, texture: {}});
+    }
+    // Uniform buffer binding layout.
+    bindGroupLayoutEntries.push(
+        {binding: 2, visibility: GPUShaderStage.COMPUTE, buffer: {}});
+    const fromPixelBindGroupLayout =
+        this.device.createBindGroupLayout({entries: bindGroupLayoutEntries});
+    const fromPixelPipelineLayout = this.device.createPipelineLayout(
+        {bindGroupLayouts: [fromPixelBindGroupLayout]});
+    return {
+      bindGroupLayout: fromPixelBindGroupLayout,
+      pipelineLayout: fromPixelPipelineLayout
+    };
+  }
+
+  private copyExternalImageToTexture(
+      externalImage: ExternalImage|HTMLVideoElement,
+      outShape: number[]): GPUTextureView {
+    const textureUsage = GPUTextureUsage.COPY_DST |
+        GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING;
+    const textureFormat = 'rgba8unorm' as GPUTextureFormat;
+    const texture = this.textureManager.acquireTexture(
+        outShape[1], outShape[0], textureFormat, textureUsage);
+    const externalResource = texture.createView();
+
+    this.queue.copyExternalImageToTexture(
+        {source: externalImage as ExternalImage, origin: {x: 0, y: 0}},
+        {texture}, [outShape[1], outShape[0]]);
+
+    const textureInfo = {
+      width: outShape[1],
+      height: outShape[0],
+      format: textureFormat,
+      usage: textureUsage,
+      texture
+    };
+    this.textureDisposalQueue.push(textureInfo);
+    return externalResource;
+  }
+
   runFromPixelsProgram(
-      program: FromPixelsProgram, output: GPUBuffer, layout: WebGPULayout,
-      externalResource: GPUExternalTexture|GPUTextureView, outputId: DataId) {
+      program: webgpu_program.WebGPUProgram, output: TensorInfo,
+      programUniforms: ProgramUniform, useImport: boolean,
+      externalImage: ExternalImage|HTMLVideoElement) {
     program.dispatch = reshapeDispatch(this.device, program);
+
+    const outputTypes = [output.dtype, useImport ? 'import' : 'copyExternal'];
+    const key =
+        webgpu_program.makeShaderKey(program, [output.shape], outputTypes);
+
+    const layout = this.getFromPixelTextureLayout(useImport);
+
+    const pipeline = this.getAndSavePipeline(key, () => {
+      return webgpu_program.compileProgram(
+          this.device, program, layout.pipelineLayout, [], output, true);
+    });
+
+    let externalResource: GPUExternalTexture|GPUTextureView;
+    if (useImport) {
+      const externalTextureDescriptor = {
+        source: externalImage as HTMLVideoElement
+      };
+      externalResource =
+          this.device.importExternalTexture(externalTextureDescriptor);
+    } else {
+      externalResource =
+          this.copyExternalImageToTexture(externalImage, output.shape);
+    }
+
+    const info = this.tensorMap.get(output.dataId);
+
+    info.bufferInfo.buffer = this.acquireBuffer(info.bufferInfo.byteSize);
```

Comment:
Use `uploadToGPU`? And maybe like `runWebGPUProgram`, we return the output tensor at last?

Reply:
done